docs/langref/relay_expr.rst

```diff
@@ -0,0 +1,396 @@
+appears or the :code:`let` expression where it is bound, respectively.
+
+Suppose the local variable :code:`%a` has been defined in a scope
+and within that scope, a :code:`let` expression binding to a variable
+:code:`%a` appears. This is permitted, as in most functional languages.
+In the scope of the :code:`let` expression (the inner scope),
+the name :code:`%a` is "shadowed," meaning all references to
+:code:`%a` in the inner scope refer to the later defintion, while
+references to :code:`%a` in the outer scope continue to refer to
+the previous one.
+
+(Note that in Relay's implemetnation, each definition of a local variable
+creates a new :py:class:`~tvm.relay.expr.Var`, so a shadowed local variable,
+despite having the same name as one in an outer scope, will be a different
+object. This allows for comparing local variables by pointer identity with the
+knowledge that the same local variable object corresponds to a different binding site.)
+
+See :py:class:`~tvm.relay.expr.Var` for its implementation
+and documentation.
+
+Global Functions
+================
+
+A function definition consists of a name, arguments, return type,
+type parameters, and any applicable type relations.
+A function's return type and parameter types may be omitted;
+Relay will attempt to infer the most general types where types
+are omitted.
+
+Functions in Relay act similarly to procedures or functions in
+other programming language and serve to generalize the concept
+of a named subgraph.
+
+Functions defined in the manner described in this subsection are
+of global scope; anonymous functions will be discussed later, though
+their mechanics are nearly identical. Note that global functions may be
+recursive; that is, within the function's body, the function's
+identifier refers back to the function unless it is shadowed in a :code:`let`
+expression.
+
+A definition minimally consists of an identifier :code:`@id`, an empty set of
+parameters, and a body expression (:py:class:`~tvm.relay.expr.Expr`)
+contained by curly braces.
+
+.. code-block:: python
+
+    def @id() { body }
+
+A definition may contain any number of parameters. For example, a
+simple function that invokes the `add` operator:
+
+.. code-block:: python
+
+    def @plus(%x, %y) { add(%x, %y) }
+
+Notice that within the function's body, the parameters are local
+variables, just like those bound in a :code:`let` expression.
+
+One may also annotate explicit types on definitions.
+For example, we can restrict the above definition to only work
+on certain types:
+
+.. code-block:: python
+
+    def @plus(%x: Tensor<(10, 10), float32>, %y: Tensor<(10, 10), float32>)
+               -> Tensor<(10, 10), float32> {
+        add(%x, %y)
+    }
+
+A function parameter is just a local variable (:py:class:`~tvm.relay.expr.LocalVar`) optionally
+annotated with a type. Parameters are written as :code:`%x : T`.
+
+When the type information is omitted, we will attempt to infer the most general type
```

Comment:
These two paragraphs need some love. The first repeats itself. More importantly, the two paragraphs don't agree on the audience's background. The first paragraph explains generalization and type inference. The second paragraph assumes the reader is familiar with type polymorphism and type parameters.

Reply:
I took a stab at making the description more explicit; please review

Reply:
lgtm